File: go/attribution/v3_validate.go
package attribution

import (
	"bytes"
	_ "embed"
	"encoding/json"
	"fmt"
	"regexp"
	"strconv"
	"strings"
	"time"

	"github.com/santhosh-tekuri/jsonschema/v5"
	"github.com/shopspring/decimal"
)

//go:embed attribution-v3-schema.json
var attributionV3SchemaJSON string

var attributionV3Schema *jsonschema.Schema

var (
	v3TimestampPattern = regexp.MustCompile(`^\d{4}-\d{2}-\d{2}T\d{2}:\d{2}:\d{2}(?:\.\d{1,6})?(?:Z|[+-]\d{2}:\d{2})$`)
	v3TimeMetrics      = stringSet("audio_seconds", "connected_seconds", "recording_seconds", "agent_seconds", "compute_seconds", "vcpu_seconds", "memory_gib_seconds", "gpu_seconds")
)

func init() {
	compiler := jsonschema.NewCompiler()
	compiler.Draft = jsonschema.Draft2020
	const schemaID = "https://schemas.dexcost.io/attribution/conformance-v3.json"
	// Go's regexp engine has no lookahead. Remove only the positivity lookahead
	// from the in-memory compiler copy; semantic validation below enforces > 0.
	schemaForCompiler := strings.ReplaceAll(attributionV3SchemaJSON, "(?=.*[1-9])", "")
	if err := compiler.AddResource(schemaID, strings.NewReader(schemaForCompiler)); err != nil {
		panic(fmt.Errorf("dexcost: load attribution v3 schema: %w", err))
	}
	var err error
	attributionV3Schema, err = compiler.Compile(schemaID + "#/components/schemas/AttributionObservation")
	if err != nil {
		panic(fmt.Errorf("dexcost: compile attribution v3 observation schema: %w", err))
	}
}

// ValidateObservationV3 validates the complete published schema and the
// control-plane cross-field invariants that JSON Schema cannot express.
func ValidateObservationV3(value interface{}) ValidationResult {
	issues := make([]ValidationIssue, 0)
	add := func(path, message string) {
		for _, issue := range issues {
			if issue.Path == path && issue.Message == message {
				return
			}
		}
		issues = append(issues, ValidationIssue{Path: path, Message: message})
	}

	var normalized interface{}
	encoded, err := json.Marshal(value)
	if err == nil {
		decoder := json.NewDecoder(bytes.NewReader(encoded))
		decoder.UseNumber()
		err = decoder.Decode(&normalized)
	}
	if err != nil {
		add("", "Observation must be JSON serializable")
		return ValidationResult{Success: false, Issues: issues}
	}

	if err := attributionV3Schema.Validate(normalized); err != nil {
		if validation, ok := err.(*jsonschema.ValidationError); ok {
			collectV3SchemaIssues(validation, add)
		} else {
			add("", err.Error())
		}
	}
	if event, ok := normalized.(map[string]interface{}); ok {
		validateV3Semantics(event, add)
	}
	return ValidationResult{Success: len(issues) == 0, Issues: issues}
}

func collectV3SchemaIssues(err *jsonschema.ValidationError, add func(string, string)) {
	if len(err.Causes) == 0 {
		path := jsonPointerPath(err.InstanceLocation)
		if path == "" && strings.HasPrefix(err.Message, "additionalProperties '") {
			parts := strings.Split(err.Message, "'")
			if len(parts) >= 2 {
				path = parts[1]
			}
		}
		add(path, err.Message)
		return
	}
	for _, cause := range err.Causes {
		collectV3SchemaIssues(cause, add)
	}
}

func jsonPointerPath(pointer string) string {
	pointer = strings.TrimPrefix(pointer, "/")
	if pointer == "" {
		return ""
	}
	parts := strings.Split(pointer, "/")
	for i := range parts {
		parts[i] = strings.ReplaceAll(strings.ReplaceAll(parts[i], "~1", "/"), "~0", "~")
	}
	return strings.Join(parts, ".")
}

func validateV3Semantics(event map[string]interface{}, add func(string, string)) {
	validV3Timestamp(event["occurred_at"], "occurred_at", add)
	validV3Timestamp(event["observed_at"], "observed_at", add)

	period, _ := event["usage_period"].(map[string]interface{})
	if period != nil {
		start, startOK := validV3Timestamp(period["start_at"], "usage_period.start_at", add)
		if rawEnd, exists := period["end_at"]; exists {
			end, endOK := validV3Timestamp(rawEnd, "usage_period.end_at", add)
			if startOK && endOK && end.Before(start) {
				add("usage_period.end_at", "Cannot precede start_at")
			}
		}
	}

	usage, _ := event["usage"].([]interface{})
	lineIDs := make(map[string]struct{}, len(usage))
	hasTimeMetric := false
	for index, raw := range usage {
		line, ok := raw.(map[string]interface{})
		if !ok {
			continue
		}
		quantity, quantityOK := line["quantity"].(string)
		parsedQuantity, quantityErr := decimal.NewFromString(quantity)
		if !quantityOK || quantityErr != nil || !parsedQuantity.IsPositive() {
			add(fmt.Sprintf("usage.%d.quantity", index), "Must be a positive plain decimal")
		}
		if lineID, ok := line["line_id"].(string); ok {
			if _, exists := lineIDs[lineID]; exists {
				add(fmt.Sprintf("usage.%d.line_id", index), "Must be unique in a full snapshot")
			}
			lineIDs[lineID] = struct{}{}
		}
		metric, _ := line["metric"].(string)
		if unit, known := UnitByMetric[UsageMetric(metric)]; known {
			if line["unit"] != string(unit) {
				add(fmt.Sprintf("usage.%d.unit", index), "Must use the canonical unit")
			}
			if v3TimeMetrics[metric] {
				hasTimeMetric = true
			}
		}
		dimensions, _ := line["dimensions"].([]interface{})
		keys := make(map[string]struct{}, len(dimensions))
		for dimensionIndex, rawDimension := range dimensions {
			dimension, ok := rawDimension.(map[string]interface{})
			if !ok {
				continue
			}
			if key, ok := dimension["key"].(string); ok {
				if _, exists := keys[key]; exists {
					add(fmt.Sprintf("usage.%d.dimensions.%d.key", index, dimensionIndex), "Must be unique within the usage line")
				}
				keys[key] = struct{}{}
			}
		}
	}

	operation, _ := event["operation"].(map[string]interface{})
	attempt, _ := operation["attempt"].(map[string]interface{})
	attemptNumber, attemptNumberOK := jsonInteger(attempt["number"])
	_, hasRetryOf := attempt["retry_of"]
	if attemptNumberOK && attemptNumber == 1 && hasRetryOf {
		add("operation.attempt.retry_of", "Attempt 1 cannot retry another attempt")
	}
	if attemptNumberOK && attemptNumber > 1 && !hasRetryOf {
		add("operation.attempt.retry_of", "Later attempts require retry_of")
	}
	attemptID, attemptIDOK := attempt["id"].(string)
	retryOf, retryOfOK := attempt["retry_of"].(string)
	if attemptIDOK && retryOfOK && attemptID == retryOf {
		add("operation.attempt.retry_of", "Attempt cannot retry itself")
	}
	if operation["status"] == "succeeded" && operation["error"] != nil {
		add("operation.error", "A succeeded operation cannot carry an error")
	}

	capability, _ := event["capability"].(map[string]interface{})
	if capability != nil && capability["source_id"] != nil && capability["source"] == nil {
		add("capability.source_id", "source_id requires source")
	}

	lifecycle, _ := event["lifecycle"].(map[string]interface{})
	state, _ := lifecycle["state"].(string)
	costEvidence, _ := event["cost_evidence"].(map[string]interface{})
	_, hasCostEvidence := event["cost_evidence"]
	hasClosedPeriod := period != nil && period["end_at"] != nil
	switch state {
	case "pending":
		if len(usage) != 0 {
			add("usage", "Pending cannot assert usage")
		}
		if hasCostEvidence {
			add("cost_evidence", "Pending cannot assert cost evidence")
		}
		if hasClosedPeriod {
			add("usage_period.end_at", "Pending cannot close usage")
		}
	case "provisional":
		if len(usage) == 0 {
			add("usage", "Provisional requires usage")
		}
		if costEvidence["confidence"] == "exact" {
			add("cost_evidence.confidence", "Provisional cost cannot be exact")
		}
	case "final":
		if operation["status"] == "in_progress" {
			add("operation.status", "Final operation cannot be in progress")
		}
		if operation["status"] == "succeeded" && len(usage) == 0 {
			add("usage", "Successful final operation requires usage")
		}
	case "voided":
		revision, ok := jsonInteger(lifecycle["revision"])
		if !ok || revision <= 1 {
			add("lifecycle.revision", "Voided revision must exceed 1")
		}
		if len(usage) != 0 {
			add("usage", "Voided cannot assert usage")
		}
		if hasCostEvidence {
			add("cost_evidence", "Voided cannot assert cost evidence")
		}
	}
	if (state == "provisional" || state == "final") && hasTimeMetric && !hasClosedPeriod {
		add("usage_period.end_at", "Time-based usage requires a closed period")
	}

	if hasCostEvidence {
		amount, amountOK := costEvidence["amount"].(string)
		parsedAmount, amountErr := decimal.NewFromString(amount)
		if !amountOK || amountErr != nil || !parsedAmount.IsPositive() {
			add("cost_evidence.amount", "Must be a positive plain decimal")
		}
	}

	source, _ := costEvidence["source"].(string)
	confidence, _ := costEvidence["confidence"].(string)
	if source == "provider_reported" && confidence != "exact" && confidence != "estimated" {
		add("cost_evidence.confidence", "Provider-reported evidence must be exact or estimated")
	}
	if source == "sdk_catalog" || source == "sdk_rate_registry" {
		if confidence == "exact" {
			add("cost_evidence.confidence", "SDK evidence cannot be exact")
		}
		version, _ := costEvidence["pricing_version"].(string)
		if version == "" {
			add("cost_evidence.pricing_version", "SDK evidence requires a pricing version")
		}
	}
}

func validV3Timestamp(value interface{}, path string, add func(string, string)) (time.Time, bool) {
	raw, ok := value.(string)
	if !ok || !v3TimestampPattern.MatchString(raw) {
		add(path, "Must be a valid offset-aware ISO 8601 instant")
		return time.Time{}, false
	}
	parsed, err := time.Parse(time.RFC3339Nano, raw)
	if err != nil {
		add(path, "Must be a valid offset-aware ISO 8601 instant")
		return time.Time{}, false
	}
	return parsed, true
}

func jsonInteger(value interface{}) (int64, bool) {
	switch number := value.(type) {
	case json.Number:
		parsed, err := strconv.ParseInt(number.String(), 10, 64)
		return parsed, err == nil
	case float64:
		if number != float64(int64(number)) {
			return 0, false
		}
		return int64(number), true
	case int:
		return int64(number), true
	case int64:
		return number, true
	default:
		return 0, false
	}
}

File: typescript/src/attribution/v3-validate.ts
import { readFileSync } from "node:fs";
import { dirname, join } from "node:path";
import { createRequire } from "node:module";
import { fileURLToPath } from "node:url";
import type { ErrorObject, ValidateFunction } from "ajv";

import type { AttributionObservationV3 } from "./v3-types.js";

export interface AttributionV3ValidationIssue {
  path: string;
  message: string;
}

export interface AttributionV3ValidationResult {
  success: boolean;
  issues: AttributionV3ValidationIssue[];
}

const schemaPath = join(
  dirname(fileURLToPath(import.meta.url)),
  "attribution-v3-schema.json",
);
const schemaDocument = JSON.parse(readFileSync(schemaPath, "utf8")) as {
  $id: string;
};
type AjvRuntime = {
  addSchema(schema: unknown): unknown;
  getSchema(key: string): ValidateFunction | undefined;
};
type AjvConstructor = new (options: Record<string, unknown>) => AjvRuntime;
type AddFormats = (instance: AjvRuntime) => unknown;
const runtimeRequire = createRequire(import.meta.url);
const AjvModule = runtimeRequire("ajv/dist/2020") as unknown;
const AjvClass = ((AjvModule as { default?: unknown }).default ?? AjvModule) as AjvConstructor;
const formatsModule = runtimeRequire("ajv-formats") as unknown;
const installFormats = ((formatsModule as { default?: unknown }).default ?? formatsModule) as AddFormats;
const ajv = new AjvClass({ allErrors: true, strict: false, discriminator: true });
installFormats(ajv);
ajv.addSchema(schemaDocument);
const compiledSchemaValidator = ajv.getSchema(
  `${schemaDocument.$id}#/components/schemas/AttributionObservation`,
);
if (compiledSchemaValidator === undefined) {
  throw new Error("DexCost v3 observation schema is missing from the SDK package");
}
const schemaValidator: ValidateFunction = compiledSchemaValidator;

const KNOWN_UNIT_BY_METRIC: Readonly<Record<string, string>> = Object.freeze({
  input_tokens: "Tokens",
  input_image_tokens: "Tokens",
  output_image_tokens: "Tokens",
  output_tokens: "Tokens",
  cache_read_input_tokens: "Tokens",
  cache_write_input_tokens: "Tokens",
  reasoning_output_tokens: "Tokens",
  characters: "Characters",
  audio_seconds: "Seconds",
  connected_seconds: "Seconds",
  recording_seconds: "Seconds",
  agent_seconds: "Seconds",
  compute_seconds: "Seconds",
  vcpu_seconds: "vCPU-Seconds",
  memory_gib_seconds: "GiB-Seconds",
  gpu_seconds: "GPU-Seconds",
  request_count: "Requests",
  call_count: "Calls",
  bytes_in: "Bytes",
  bytes_out: "Bytes",
  image_count: "Images",
  page_count: "Pages",
  credit_count: "Credits",
});

const TIME_METRICS = new Set([
  "audio_seconds",
  "connected_seconds",
  "recording_seconds",
  "agent_seconds",
  "compute_seconds",
  "vcpu_seconds",
  "memory_gib_seconds",
  "gpu_seconds",
]);

const TIMESTAMP = /^(\d{4})-(\d{2})-(\d{2})T(\d{2}):(\d{2}):(\d{2})(?:\.(\d{1,6}))?(?:Z|[+-](\d{2}):(\d{2}))$/;

function isRecord(value: unknown): value is Record<string, unknown> {
  return value !== null && typeof value === "object" && !Array.isArray(value);
}

function schemaErrorPath(error: ErrorObject): string {
  const parts = error.instancePath === ""
    ? []
    : error.instancePath
      .slice(1)
      .split("/")
      .map((part) => part.replaceAll("~1", "/").replaceAll("~0", "~"));
  if (error.keyword === "required") {
    parts.push(String((error.params as { missingProperty?: unknown }).missingProperty ?? ""));
  }
  if (error.keyword === "additionalProperties") {
    parts.push(String((error.params as { additionalProperty?: unknown }).additionalProperty ?? ""));
  }
  return parts.filter((part) => part !== "").join(".");
}

function validCalendarTimestamp(value: string): RegExpExecArray | undefined {
  const match = TIMESTAMP.exec(value);
  if (match === null) return undefined;
  const year = Number(match[1]);
  const month = Number(match[2]);
  const day = Number(match[3]);
  const hour = Number(match[4]);
  const minute = Number(match[5]);
  const second = Number(match[6]);
  const offsetHour = match[8] === undefined ? 0 : Number(match[8]);
  const offsetMinute = match[9] === undefined ? 0 : Number(match[9]);
  const leapYear = year % 4 === 0 && (year % 100 !== 0 || year % 400 === 0);
  const daysInMonth = [31, leapYear ? 29 : 28, 31, 30, 31, 30, 31, 31, 30, 31, 30, 31];
  if (
    month < 1 || month > 12 || day < 1 || day > daysInMonth[month - 1] ||
    hour > 23 || minute > 59 || second > 59 || offsetHour > 23 || offsetMinute > 59 ||
    !Number.isFinite(Date.parse(value))
  ) {
    return undefined;
  }
  return match;
}

function timestampMicroseconds(value: unknown): bigint | undefined {
  if (typeof value !== "string") return undefined;
  const match = validCalendarTimestamp(value);
  if (match === undefined) return undefined;
  const milliseconds = Date.parse(value);
  const fractional = (match[7] ?? "").padEnd(6, "0");
  return BigInt(milliseconds) * 1_000n + BigInt(fractional.slice(3) || "0");
}

function addTimestampIssue(
  value: unknown,
  path: string,
  issues: AttributionV3ValidationIssue[],
): void {
  if (timestampMicroseconds(value) === undefined) {
    issues.push({ path, message: "Must be a valid offset-aware ISO 8601 instant" });
  }
}

function semanticIssues(value: unknown): AttributionV3ValidationIssue[] {
  if (!isRecord(value)) return [];
  const issues: AttributionV3ValidationIssue[] = [];
  addTimestampIssue(value.occurred_at, "occurred_at", issues);
  addTimestampIssue(value.observed_at, "observed_at", issues);

  const usagePeriod = isRecord(value.usage_period) ? value.usage_period : undefined;
  if (usagePeriod !== undefined) {
    addTimestampIssue(usagePeriod.start_at, "usage_period.start_at", issues);
    if (usagePeriod.end_at !== undefined) {
      addTimestampIssue(usagePeriod.end_at, "usage_period.end_at", issues);
      const start = timestampMicroseconds(usagePeriod.start_at);
      const end = timestampMicroseconds(usagePeriod.end_at);
      if (start !== undefined && end !== undefined && end < start) {
        issues.push({ path: "usage_period.end_at", message: "Cannot precede start_at" });
      }
    }
  }

  const usage = Array.isArray(value.usage) ? value.usage : [];
  const lineIds = new Set<unknown>();
  for (const [lineIndex, rawLine] of usage.entries()) {
    if (!isRecord(rawLine)) continue;
    if (lineIds.has(rawLine.line_id)) {
      issues.push({
        path: `usage.${lineIndex}.line_id`,
        message: "Must be unique in a full snapshot",
      });
    }
    lineIds.add(rawLine.line_id);
    const metric = typeof rawLine.metric === "string" ? rawLine.metric : undefined;
    const canonicalUnit = metric === undefined ? undefined : KNOWN_UNIT_BY_METRIC[metric];
    if (canonicalUnit !== undefined && rawLine.unit !== canonicalUnit) {
      issues.push({ path: `usage.${lineIndex}.unit`, message: `Must be ${canonicalUnit}` });
    }
    const dimensions = Array.isArray(rawLine.dimensions) ? rawLine.dimensions : [];
    const dimensionKeys = new Set<unknown>();
    for (const [dimensionIndex, rawDimension] of dimensions.entries()) {
      if (!isRecord(rawDimension)) continue;
      if (dimensionKeys.has(rawDimension.key)) {
        issues.push({
          path: `usage.${lineIndex}.dimensions.${dimensionIndex}.key`,
          message: "Must be unique within the usage line",
        });
      }
      dimensionKeys.add(rawDimension.key);
    }
  }

  const operation = isRecord(value.operation) ? value.operation : undefined;
  const attempt = operation !== undefined && isRecord(operation.attempt)
    ? operation.attempt
    : undefined;
  if (attempt?.number === 1 && attempt.retry_of !== undefined) {
    issues.push({
      path: "operation.attempt.retry_of",
      message: "Attempt 1 cannot retry another attempt",
    });
  }
  if (typeof attempt?.number === "number" && attempt.number > 1 && attempt.retry_of === undefined) {
    issues.push({
      path: "operation.attempt.retry_of",
      message: "Later attempts require retry_of",
    });
  }
  if (attempt?.id !== undefined && attempt.id === attempt.retry_of) {
    issues.push({ path: "operation.attempt.retry_of", message: "Attempt cannot retry itself" });
  }
  if (operation?.status === "succeeded" && operation.error !== undefined) {
    issues.push({ path: "operation.error", message: "A succeeded operation cannot carry an error" });
  }

  const capability = isRecord(value.capability) ? value.capability : undefined;
  if (capability?.source_id !== undefined && capability.source === undefined) {
    issues.push({ path: "capability.source_id", message: "source_id requires source" });
  }

  const lifecycle = isRecord(value.lifecycle) ? value.lifecycle : undefined;
  const state = lifecycle?.state;
  const costEvidence = isRecord(value.cost_evidence) ? value.cost_evidence : undefined;
  if (state === "pending") {
    if (usage.length !== 0) issues.push({ path: "usage", message: "Pending cannot assert usage" });
    if (value.cost_evidence !== undefined) {
      issues.push({ path: "cost_evidence", message: "Pending cannot assert cost evidence" });
    }
    if (usagePeriod?.end_at !== undefined) {
      issues.push({ path: "usage_period.end_at", message: "Pending cannot close usage" });
    }
  } else if (state === "provisional") {
    if (usage.length === 0) issues.push({ path: "usage", message: "Provisional requires usage" });
    if (costEvidence?.confidence === "exact") {
      issues.push({ path: "cost_evidence.confidence", message: "Provisional cost cannot be exact" });
    }
  } else if (state === "final") {
    if (operation?.status === "in_progress") {
      issues.push({ path: "operation.status", message: "Final operation cannot be in progress" });
    }
    if (operation?.status === "succeeded" && usage.length === 0) {
      issues.push({ path: "usage", message: "Successful final operation requires usage" });
    }
  } else if (state === "voided") {
    if (typeof lifecycle?.revision !== "number" || lifecycle.revision <= 1) {
      issues.push({ path: "lifecycle.revision", message: "Voided revision must exceed 1" });
    }
    if (usage.length !== 0) issues.push({ path: "usage", message: "Voided cannot assert usage" });
    if (value.cost_evidence !== undefined) {
      issues.push({ path: "cost_evidence", message: "Voided cannot assert cost evidence" });
    }
  }

  if (
    (state === "provisional" || state === "final") &&
    usage.some((rawLine) => isRecord(rawLine) && TIME_METRICS.has(String(rawLine.metric))) &&
    usagePeriod?.end_at === undefined
  ) {
    issues.push({
      path: "usage_period.end_at",
      message: "Time-based usage requires a closed period",
    });
  }

  if (costEvidence?.source === "provider_reported" &&
      costEvidence.confidence !== "exact" && costEvidence.confidence !== "estimated") {
    issues.push({
      path: "cost_evidence.confidence",
      message: "Provider-reported evidence must be exact or estimated",
    });
  }
  if (costEvidence?.source === "sdk_catalog" || costEvidence?.source === "sdk_rate_registry") {
    if (costEvidence.confidence === "exact") {
      issues.push({ path: "cost_evidence.confidence", message: "SDK evidence cannot be exact" });
    }
    if (typeof costEvidence.pricing_version !== "string" || costEvidence.pricing_version.length === 0) {
      issues.push({
        path: "cost_evidence.pricing_version",
        message: "SDK evidence requires a pricing version",
      });
    }
  }
  return issues;
}

/** Validate the complete v3 schema plus control-plane cross-field invariants. */
export function validateAttributionObservationV3(value: unknown): AttributionV3ValidationResult {
  const issues: AttributionV3ValidationIssue[] = [];
  if (!schemaValidator(value)) {
    for (const error of schemaValidator.errors ?? []) {
      issues.push({
        path: schemaErrorPath(error),
        message: error.message ?? "Schema validation failed",
      });
    }
  }
  issues.push(...semanticIssues(value));
  const unique = issues.filter(
    (issue, index, all) => all.findIndex(
      (candidate) => candidate.path === issue.path && candidate.message === issue.message,
    ) === index,
  );
  return { success: unique.length === 0, issues: unique };
}

export function assertAttributionObservationV3(
  value: unknown,
): asserts value is AttributionObservationV3 {
  const result = validateAttributionObservationV3(value);
  if (!result.success) {
    throw new Error(result.issues.map((issue) => `${issue.path}: ${issue.message}`).join("; "));
  }
}
